### DISCUSSION TITLE
saveMenuItem does not save the magic key by design

### DISCUSSION THREAD

**Discussion by luwenjin** [davetcc/tcMenu#299] [Q&A]
How to reproduce: 
1. flash a new rom to a new MCU
2. edit menu value, `saveMenuItem + EEPROM.commit()`, reset;  **the value of the menuItem is not saved** 
4. run `menuMgr.save();EEPROM.commit();`, reset; the value of the menu is empty as before
5. edit menu value, run the same `saveMenuItem + EEPROM.commit()` again, reset; the value is saved

I guess it's because saveMenuItem does't write magic to the EEPROM?

**Comment by davetcc** [comment]
I've converted this into a discussion, so it doesn't get closed and out of view.

This is by design, in order for the menu to be saved into a safe state, suitable for loading back by load, the whole menu must be saved the first time around. If this were not true, it would be possible to crash the board by loading back an unterminated string, or bring a system up with menus in a partially corrupt state.

If you wanted to override this behaviour in your design, and you accept the risk of what's in the ROM, you could yourself write out the header information (IE the magic key and the size field - but what would you set that size field to in that case?)

**Comment by davetcc** [comment]
If you decided it was a good idea to write the key, and I wouldn't personally recommend doing so, you'd just get the EEPROM abstraction from menu manager, and write the extra two fields yourself, something along the lines of.

```
auto eepromPtr = menuMgr.getEepromAbstraction();
eepromPtr->write16(2, myMagicKey);
eepromPtr->write16(4, whateverSizeIsSafe);
```

**Reply by luwenjin** [reply]
Thanks!
I see the reason behind the design.
I decided to save the whole menu at the first time launch to solve the problem.
